src/GraphQL/Validation/ValidationContext.cs

```diff
@@ -20,6 +21,11 @@ public class ValidationContext : IProvideUserContext
         private readonly Dictionary<Operation, IEnumerable<VariableUsage>> _variables =
             new Dictionary<Operation, IEnumerable<VariableUsage>>();
 
+        /// <summary>
+        /// Allows validation rules store their specific data during validation.
+        /// </summary>
+        private readonly ConcurrentDictionary<object, object> _validationRuleLocalContext = new ConcurrentDictionary<object, object>();
```

Comment:
Since `Get<TValidationRule, TResult>` and `Set<>` are `internal` and not public, there is no need for a concurrent dictionary.

Reply:
Why? Rules are executed in parallel by `DocumentValidator`.

Reply:
No they are not.  They "asynchronously" retrieve the `INodeVisitor` instances, and then run them synchronously in the `BasicVisitor`.  But since none of the rules have an `await` in them, they will all return the `INodeVisitor` instances synchronously also.  It's 100% synchronous with the default rule set (whether its the old code in develop, or any of our branches).

The only thing that could be asynchronous is if there was a custom validation rule that, _before_ returning the `INodeVisitor`, awaited another method -- that portion of that one validation rule's code would be asynchronous.  But that is not part of the core rule set that has access to this dictionary.

Reply:
🤦 🤦 🤦  I remember thinking about it, but then something distracted me and I forgot.